From oliver@neukum.org Mon Sep 17 11:15:16 2007
From: Oliver Neukum <oliver@neukum.org>
Date: Mon, 17 Sep 2007 20:15:53 +0200
Subject: USB: fix double frees in error code paths of ipaq driver
To: Alan Stern <stern@rowland.harvard.edu>, "Greg Kroah-Hartman" <gregkh@suse.de>, linux-usb-devel@lists.sourceforge.net
Message-ID: <200709172015.53549.oliver@neukum.org>
Content-Disposition: inline


the error code paths can be enter with buffers to freed buffers.
Serial core would do a kfree() on memory already freed.

Signed-off-by: Oliver Neukum <oneukum@suse.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/serial/ipaq.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/usb/serial/ipaq.c
+++ b/drivers/usb/serial/ipaq.c
@@ -646,11 +646,13 @@ static int ipaq_open(struct usb_serial_p
 	kfree(port->bulk_out_buffer);
 	port->bulk_in_buffer = kmalloc(URBDATA_SIZE, GFP_KERNEL);
 	if (port->bulk_in_buffer == NULL) {
+		port->bulk_out_buffer = NULL; /* prevent double free */
 		goto enomem;
 	}
 	port->bulk_out_buffer = kmalloc(URBDATA_SIZE, GFP_KERNEL);
 	if (port->bulk_out_buffer == NULL) {
 		kfree(port->bulk_in_buffer);
+		port->bulk_in_buffer = NULL;
 		goto enomem;
 	}
 	port->read_urb->transfer_buffer = port->bulk_in_buffer;
